### DISCUSSION TITLE
phoneme based fine tuning -> how to make tacotron2 speak intelligible

### DISCUSSION THREAD

**Discussion by buczekEngineering** [coqui-ai/TTS#1525] [General Q&A]
Hi Coqui TTS Community!

I'm finetuning the german pre-trained model (from @thorstenMueller)
I have: 1286 short audio samples [22050hz, mono, 16bits, trimed_silence]
 1h audio in total, 
v.0.5.0 TTS, 
Tacotron2 DCC pre-trained model

After 40k steps the general sounds of cloned voice is similar to the target speaker, but the speech is hardly intelligible.  **It is hard to distinguish spoken words.** 
I'm also still getting the following warning while fine-tuning  **"warning: audio amplitude out of range, auto clipped."** 
How to investigate further this issue? (During the last community meeting it was recommended to ask for help if this still occurs after some time of training, so i do.)

I attach some information about the training results and data statistics as well as part of configs.json: 
Can someone have a look on it to double check if this is correctly set up? I would be very thankful for all suggestions!
---------------------------------------------
number of output frames = 2 

 | > Use phonemes: True
   | > phoneme language: de-de
 | > Number of instances : 1286
 | > Max length sequence: 166
 | > Min length sequence: 11
 | > Avg length sequence: 47.87247278382581
 | > Num. instances discarded by max-min (max=inf, min=1) seq limits: 0
 | > Batch group size: 0.

Number of unique phonemes: 76 

Unique all forced to lower phonemes:  a b d e f h i j k l m n o p r s t u v y z œ ɑ ɒ ɔ ɛ ɡ ɪ ʃ ʊ ʏ!,.;?abdefhijklmnoprstuvxyzçøŋœɐɑɔəɛɜɡɪɾʃʊʏʒː

> Avg mel spec mean: -6.109856162059318 

> Avg mel spec scale: 1.8149526576145327 

> Avg linear spec mean: -4.517870468314907 

> Avg linear spec scale: 1.7804468458203482 

Latest Evaluation Logs: 
  --> EVAL PERFORMANCE
     | > avg_loader_time: 0.00444 (+0.00003)
     | > avg_decoder_loss: 0.33067 (+0.00478)
     | > avg_postnet_loss: 0.29123 (+0.00732)
     | > avg_stopnet_loss: 0.19371 (+0.00032)
     | > avg_decoder_coarse_loss: 0.35159 (-0.00181)
     | > avg_decoder_ddc_loss: 0.01070 (-0.00004)
     | > avg_ga_loss: 0.00056 (-0.00000)
     | > avg_decoder_diff_spec_loss: 0.19582 (+0.00077)
     | > avg_postnet_diff_spec_loss: 0.15706 (+0.00066)
     | > avg_decoder_ssim_loss: 0.26963 (-0.00022)
     | > avg_postnet_ssim_loss: 0.24656 (+0.00047)
     | > avg_loss: 0.65982 (+0.00329)
     | > avg_align_error: 0.57490 (+0.00295)

**Part of config.json I am using:**
epochs": 1000,
    "batch_size": 64,
    "eval_batch_size": 16,
    "mixed_precision": false,
    "scheduler_after_epoch": false,
    "run_eval": true,
    "test_delay_epochs": -1,
    "print_eval": true,
    "dashboard_logger": "tensorboard",
    "print_step": 25,
    "plot_step": 100,
    "model_param_stats": false,
    "project_name": null,
    "log_model_step": null,
    "wandb_entity": null,
    "save_step": 2000,
    "checkpoint": true,
    "keep_all_best": true,
    "keep_after": 10000,
    "num_loader_workers": 4,
    "num_eval_loader_workers": 4,
    "use_noise_augment": false,
    "use_language_weighted_sampler": false,
"audio": {
        "fft_size": 1024,
        "win_length": 1024,
        "hop_length": 256,
        "frame_shift_ms": null,
        "frame_length_ms": null,
        "stft_pad_mode": "reflect",
        "sample_rate": 22050,
        "resample": false,
        "preemphasis": 0.0,
        "ref_level_db": 20,
        "do_sound_norm": false,
        "log_func": "np.log",
        "do_trim_silence": true,
        "trim_db": 60.0,
        "do_rms_norm": false,
        "db_level": null,
        "power": 1.5,
        "griffin_lim_iters": 60,
        "num_mels": 80,
        "mel_fmin": 50,
        "mel_fmax": null,
        "spec_gain": 1.0,
        "do_amp_to_db_linear": true,
        "do_amp_to_db_mel": true,
        "signal_norm": false,
        "min_level_db": -100,
        "symmetric_norm": true,
        "max_norm": 4.0,
        "clip_norm": true,
        "stats_path": null
    },
    "use_phonemes": true,
    "use_espeak_phonemes": true,
    "phoneme_language": "de-de",
    "compute_input_seq_cache": false,
    "text_cleaner": "phoneme_cleaners",
    "enable_eos_bos_chars": false,
    "test_sentences_file": "",
    "phoneme_cache_path": "/home/.path/to/phoneme_cache",
    "characters": {
        "pad": "_",
        "eos": "~",
        "bos": "^",
        "characters": "ABCDEFGHIJKLMNOPQRSTUVWXYZ!',-.:;?abcdefghijklmnopqrstuvwxyz\u00df\u00e4\u00e9\u00f6\u00fc\u0308\u2012\u2013\u2014\u2018\u2019\u201c\u201e ",
        "punctuations": "!'(),-.:;? \u2012\u2013\u2014\u2018\u2019",
        "phonemes": " a b d e f h i j k l m n o p r s t u v w y z \u00e7 \u00f0 \u00f8 \u0153 \u0251 \u0252 \u0254 \u025b \u0261 \u026a \u0279 \u0283 \u028a \u028c \u028f!!!!!!!?,....:;??!abdefhijklmnoprstuvwxyz\u00e7\u00f8\u014b\u0153\u0250\u0251\u0252\u0254\u0259\u025b\u025c\u0261\u026a\u0279\u027e\u0283\u028a\u028c\u028f\u0292\u02d0\u0329\u0303",
        "unique": true
    },
    "batch_group_size": 0,
    "loss_masking": true,
    "sort_by_audio_len": false,
    "min_seq_len": 1,
    "max_seq_len": Infinity,
    "compute_f0": false,
    "compute_linear_spec": false,
    "add_blank": false,
    "datasets": [
        {
            "name": "ljspeech",
            "path": "dataset_final_22050/",
            "meta_file_train": "metadata.csv",
            "ignored_speakers": null,
            "language": "de-de",
            "meta_file_val": "metadata.csv",
            "meta_file_attn_mask": ""
        }
    ],
    "optimizer": "RAdam",
    "optimizer_params": {
        "betas": [
            0.9,
            0.998
        ],
        "weight_decay": 1e-06
    },
    "lr_scheduler": "NoamLR",
    "lr_scheduler_params": {
        "warmup_steps": 4000
    },
    "test_sentences": [
        "Und \u00fcberzeugen dank feingef\u00fchl f\u00fcr den ganz gro\u00dfen leinwand-stoff.",
        "Zur schadensh\u00f6he gab es keine angaben.",
        "Au\u00dferdem k\u00f6nnen glasscheiben, w\u00e4nde und andere hindernisse das ergebnis beeinflussen.",
        "Ihre lippen m\u00fcssen dennoch nicht zwingend auf farbe verzichten.",
        "Es dauert lange, eine eigene Stimme zu entwickeln, aber jetzt wo ich sie habe, bin ich nie wieder still.",
        "Heute scheint die Sonne, Regen ist nicht zu erwarten.",
        "Die aktuelle Au\u00dfentemperatur betr\u00e4gt zw\u00f6lf Grad Celsius bei einer Luftfeuchtigkeit von achtunddrei\u00dfig Prozent."
    ],
    "use_gst": false,
    "gst": null,
    "gst_style_input": null,
    "num_speakers": 1,
    "num_chars": 73,
    "r": 6,
    "gradual_training": [
        [
            0,
            6,
            64
        ],
        [
            330000,
            4,
            32
        ],
        [
            360000,
            3,
            32
        ],
        [
            370000,
            2,
            32
        ]
    ],
    "memory_size": -1,
    "prenet_type": "original",
    "prenet_dropout": true,
    "prenet_dropout_at_inference": false,
    "stopnet": true,
    "separate_stopnet": true,
    "stopnet_pos_weight": 10.0,
    "max_decoder_steps": 500,
    "encoder_in_features": 512,
    "decoder_in_features": 512,
    "decoder_output_dim": 80,
    "out_channels": 80,
    "attention_type": "original",
    "attention_heads": null,
    **"attention_norm": "softmax",**
    "attention_win": false,
    "windowing": false,
    "use_forward_attn": false,
    "forward_attn_mask": false,
    "transition_agent": false,
    "location_attn": true,
    "bidirectional_decoder": false,
    "double_decoder_consistency": true,
    "ddc_r": 6,
    "use_speaker_embedding": false,
    "speaker_embedding_dim": 512,
    "use_d_vector_file": false,
    "d_vector_file": false,
    "d_vector_dim": null,
    "lr": 0.0001,
    "grad_clip": 5.0,
    "seq_len_norm": false,
    "decoder_loss_alpha": 0.25,
    "postnet_loss_alpha": 0.25,
    "postnet_diff_spec_alpha": 0.25,
    "decoder_diff_spec_alpha": 0.25,
    "decoder_ssim_alpha": 0.25,
    "postnet_ssim_alpha": 0.25,
    "ga_alpha": 5.0
}

**Comment by p0p4k** [comment]
For now, would suggest to try out the latest repo. And train for a bit longer. Good luck. Post your results if you still face problems, then we can try to help better.

**Reply by buczekEngineering** [reply]
Ok thanks for help. I will try it out today. 
Will the model trained with v.0.5.0 be compatible with training scripts from v.0.6.x ? Or should i pick the pre-trained model from the v.0.6.x? [I know it is recommended to pick the model trained with the appropriate version, but is it going to theoretically work ?]

**Reply by p0p4k** [reply]
Pick the newest one as recommended. But you can always explore the code and see if the older model is compatible.

**Comment by buczekEngineering** [comment]
@p0p4k Sir, could you tell me if gradual reduction of "r" should be appied during finetuning? 
The @thorstenMueller tacotron2DCA model trained for 330 000 steps which currently makes the r=1 . Should I leave this parameter like this, or [since i'm using my own dataset for fine tuning] set up the gradual training starting with r=7 ..

**Reply by p0p4k** [reply]
I gave up on tacotron (not from this repo) after finding VITS. From my findings, VITS is fast to train (even from scratch) and so much faster for inference as well. Is there any particular reason you have to use tacotron?

**Reply by buczekEngineering** [reply]
@p0p4k , Sir, I need a German model. VITS is currently english, french and portuguese, right? 
Do you think it would make sense to fine-tune VITS with german 1h audio data set?

**Reply by p0p4k** [reply]
Yes, I trained a korean model using just 1 hour of data (female). Then I even fine tuned it with another 10 minutes (male) data and the results were good. Do not knock it off unless you give it a try! 
As long as there is enough phoneme coverage and a good balance of small and medium length sentences, it will work!

**Reply by buczekEngineering** [reply]
@p0p4k , Sir Did you train from scratch? or was it finetuning of one of the pre-trained Vits Models? I think I will give it a try!
 I don't know much abourt VITS, does have a vocoder already inside [real end2end]? Or did you train a vocoder additionally?

**Reply by p0p4k** [reply]
I trained from scratch. 
I used a single GPU, with mixed precision on. Resampled to 16kHz for utilising more GPU memory. 
It is real end to end. No need for training vocoder. Are you using a single speaker dataset?

**Reply by buczekEngineering** [reply]
Ok thanks for hints! I will give it a go. my Data are from a single speaker, ljspeech format. They are sampled at 22050Hz since I tried to get a match with the data from the pre-trained german model.
Did you use a recipe[train_vits.py] from the newset repo?

**Reply by buczekEngineering** [reply]
@p0p4k Sir, Vits is phoneme -based? I'm checking the config and don't see any specific flag for phoneme nor character-based training

**Reply by buczekEngineering** [reply]
@p0p4k How long did you let the VITS model train from scratch? 

Thanks a lot for you help, Sir!

**Reply by p0p4k** [reply]
I used v0.6.1. The [latest](https://github.com/coqui-ai/TTS/blob/main/recipes/ljspeech/vits_tts/train_vits.py) one has fixed some bugs in VITS. So, better that you use the latest one (not much has changed in the recipe). 
You can resample your data before training if you want to using [resample.py](https://github.com/coqui-ai/TTS/blob/c410bc58ef3bd07b72ab05d29bbdc2a6df47afea/TTS/bin/resample.py).

**Reply by p0p4k** [reply]
> @p0p4k Sir, Vits is phoneme -based? I'm checking the config and don't see any specific flag for phoneme nor character-based training

@buczekEngineering For phonemes, I think you need to use gruut as it supports `de`. Maybe espeak is okay too.
Check [this](https://github.com/coqui-ai/TTS/discussions/1490#discussioncomment-2552481) discussion from earlier.
https://github.com/coqui-ai/TTS/blob/c410bc58ef3bd07b72ab05d29bbdc2a6df47afea/recipes/ljspeech/vits_tts/train_vits.py#L46-L47

@buczekEngineering  I trained it for 550k+ steps (but results started showing up way earlier around 50k-100k steps). Read the VITS paper, and my analysis #1501 and #1494 so that you understand what you are dealing with. Also, just like you, I am a beginner and trying to experiment, make mistakes and learn more. So, no need to thank me, but we all must thank @erogol and his team for this great repo!

**Reply by domcross** [reply]
> @p0p4k , Sir, I need a German model. VITS is currently english, french and portuguese, right?
> Do you think it would make sense to fine-tune VITS with german 1h audio data set?

A german VITS model based on Thorsten-DE dataset will be released soon (with a bit of luck already in May 2022, maybe June or July).

**Reply by buczekEngineering** [reply]
Thanks guys @p0p4k @domcross !
Really looking forward to test the German Vits @thorstenMueller model! for now I will finish fientuning the Tacotron2DCA and learn more about VITS! 

[Sorry for off-topic] @domcross do you know when the German HifiGAn vocoder will be released?  Will it be compatible with Thorsten Tacotron2 DCA model from the v.0.6.1 realeses?

**Reply by domcross** [reply]
> [Sorry for off-topic] @domcross do you know when the German HifiGAn vocoder will be released? Will it be compatible with Thorsten Tacotron2 DCA model from the v.0.6.1 realeses?

We are working on final quality assurance steps and we aim to release both - Taco2DDC+HifiGAN and VITS - at the same time.
The VITS model was trained on TTS 0.6.1. Don‘t remember which version Taco2&HifiGan has been trained on…

**Reply by thorstenMueller** [reply]
I'm not @domcross but hope my answer is helpful nevertheless.
Didn't check the compatibility details but i think that HifiGAN vocoder will not be compatible with existing DCA model. We will release a VITS model and a Tacotron2 DDC with HifiGAN model.

https://user-images.githubusercontent.com/9558265/165137038-ef3f2a68-5bf5-44d7-9821-871e0731d3f4.mov

Oops, same time posting @domcross ;-).

**Reply by buczekEngineering** [reply]
@thorstenMueller @domcross the model output sounds already very good! 
I was wondering how do you guys make the model correctly pronounce abbreviations, or some foreign or custom words? I checked @thorstenMueller tutorial regarding phoneme bot, but could not apply it on the fly yet.

**Reply by thorstenMueller** [reply]
Hi @buczekEngineering ,
thanks for the feedback. Phoneme bot is still work in progress. Once the models are released i'll spend the "free" time on improving the phoneme bot. Till then i adjust the input text like it should be pronounced - so it's written incorrect but sounding better ;-).

**Comment by buczekEngineering** [comment]
I continued the fine tuning of the german tacotron2 DCC model, after reaching 60k steps of fine tuning(390k steps in general) I tested the model and recognized significant improvement in enounciation. I let the model train for the next 40k steps, though.

Problem that I still have to solve: the quality of speech drop down by longer text-sentences [ca 100chars], the further within synthesized audio, the less intelligible speech I can hear[ as i read in the documentation, it is considered as a common problem], the average sentence length in my traning dataset is ca.47 characters.  I'm wondering what could help? Any specific attention mechanism type is able to do the job? What is your experience with this particular problem @p0p4k @thorstenMueller @domcross ?  curently i'm using attention type: "original", and attention norm: "softmax".

**Reply by p0p4k** [reply]
When I used tacotron2 (from another repo), it failed for longer sentences as well. So I just split the sentences according to certain pauses (commas or periods), got the audio and then joined them with another python tool. Was not the best solution, but still worked. Waiting for other experts on their opinions. Guten tag!

**Comment by buczekEngineering** [comment]
Now I changed the value of max_decoder_steps. For fine tuning i used the value of 500, for inference I changed to 1500. Seems to be better now.

**Comment by erogol** [comment]
Since your dataset is small, you can merge it with the larger dataset and train it as a multi-speaker model. That way, you have a better chance to keep the model quality as good as the original model. 

Sorry if this is something you've already tried.

**Comment by tekinek** [comment]
My suggestions:
1) Make sure that the char/phoneme sets and their ids are exactly same in two models.
2) Keep the r value unchanged from the last r of pretrained model. I experienced worsening Stopnet performance when the r value changes during finetuning (especially with a small amount of data).
3) Finetune both text2mel and mel2wav models with the compatible audio settings.
